Include reference links in Oncoprint gene set track tooltips

fedde-s opened the pull request:
When exploring the relevance of gene sets through the heatmap functionality in the Oncoprint, users sometimes want to look up what a gene set is actually about. The metadata known to cBioPortal include the URL of the main reference for the gene set. By making the Oncoprint setup retrieve this URL via the web API, a hyperlink can be presented to the user in the track tooltip.

## Screenshot ##
![Screenshot of an Oncoprint track tooltip with a link as the header](https://user-images.githubusercontent.com/4929431/28020684-8ccf0944-6585-11e7-9725-0e397fb4d09e.png)


## Superseded screenshot of pre-review version ##
![Screenshot of an Oncoprint track tooltip with a link in the description](https://user-images.githubusercontent.com/4929431/27872112-f8161918-61a7-11e7-93b9-905bd1d20ceb.png)

fedde-s committed:
Make Oncoprint label view tooltips mouse-navigable

Add a slight delay before the tooltip disappears when moving off the
label so that the mouse can be moved onto it without it disappearing,
following the example from the cell view tooltips, which already have
links in them.

fedde-s committed:
Add web service to request gs metadata by ID list

fedde-s committed:
Include gene set links in Oncoprint tooltips

fedde-s commented:
@adamabeshouse, could you review the additions to the cBioPortal Oncoprint setup, and accomodating slight changes to the label view class? And @ersinciftci, could you take a look at the web service endpoint that I added using an existing repository method? And @pieterlukasse, could you comment on whether this is the intended functionality, in the context of the rest of the gene set visualisation effort?

adamabeshouse commented on portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js:
```diff
@@ -82,7 +84,7 @@ var OncoprintLabelView = (function () {
 		    stopDragging(view, previous_track_id);
 		}
-		view.tooltip.hide();
+		view.tooltip.hideIfNotAlreadyGoingTo(150);
```
these tooltip methods are so messy.... if you have any suggestions on how to improve that API please let me know

fedde-s replied:
I can't think of much right now. Renaming it to something like `initializeHideTimeout()` would only need *more* documentation to confirm that it wouldn't reset an already-existing timeout.

adamabeshouse reviewed (approved):
oncoprint related changes look good to me

pieterlukasse commented:
@fedde-s I found one small issue while testing. See screenshot below:

![image](https://user-images.githubusercontent.com/2900303/27904211-35be5778-623c-11e7-8aa3-3066c3a03e07.png)

Apparently the name is added once more when the name is very long?

fedde-s commented:
Yes, @pieterlukasse. This seems to come with the ellipses in the labels themselves. The boldfaced name [is inserted](https://github.com/cBioPortal/cbioportal/blob/v1.7.2/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js#L59) without whitespace to separate it from the description. Moreover, all Oncoprint tooltips [seem to be styled](https://github.com/cBioPortal/cbioportal/blob/v1.7.2/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprinttooltip.js#L5) with a maximum width of 300 of whatever the browser's fallback unit is, usually pixels, which causes the overflow since there's no space to break the line on. 

@adamabeshouse, what do you think the best solution is? I can think of two:

- make the boldfaced name a block element (`<b style="display: block;">` or `<div><b>`), to explicitly render it on a line by itself
- leave out the boldfaced name entirely if there's a description, since the description usually starts with the name anyway

In addition, the overflow of the tooltip for long single-word lines looks pretty bad, and I'm not sure what to do about that. @pieterlukasse suggests completely removing the maximum width, can you think of a better solution?

adamabeshouse commented:
@fedde-s I think the boldfaced name should definitely be a block element. I'm surprised it's not already like that, that's an oversight.

I'm not sure what the solution is as far as when the name is duplicated, e.g. for a link. I do think we should show the name in the tooltip by default, because there is not always a description. But maybe there could be an option to have the name be a link, since that's really the only reason why you would need to duplicate it.

For solving the overflow, I think for now the best solution is to have your description element have the same max width.

fedde-s committed:
Allow annotating Oncoprint tracks with a hyperlink

Add a track parameter that allows tracks to be annotated with a
hyperlink, which will appear as the first line of the track label's
tooltip if specified.

fedde-s committed:
Allow HTML versions of track names

Allow a pre-escaped HTML version of a track name to be set as a property
on String objects given as the label of a track, to include special
characters.

fedde-s committed:
Insert word breaks in gene set tooltip titles

As gene sets shown in the cBioPortal Oncoprint typically have rather
long MSigDB names containing underscores and dots, insert zero-width
spaces after these punctuation marks to allow the browser to break the
line if the name is wider than the tooltip. HTML `<wbr />` elements will
not do, as they are unsupported in later versions of Internet Explorer.

fedde-s commented:
Based on the discussion we had, I:
- made the label appear on a bold-faced line by itself (semantically a heading) if it is displayed in the tooltip;
- replaced the manual insertion of an html link into the description by a new track param named `link_url` which makes the label itself a link so that the name does not have to be repeated in the description underneath; and
- marked word breaks to allow the browser to split the line after punctuation commonly found in long MSigDB gene set names, in case the name does overflow the width of the tooltip

@adamabeshouse and @pieterlukasse, could you review the amendments?

pieterlukasse reviewed (approved):
tested on chrome and IE, with long and short names, with and without links. All works fine :+1: 

Please squash.

fedde-s commented:
@adamabeshouse, is squashing the frontend fixup commits fine with you, or do you still want to review them?

ersinciftci commented:
API part looks good to me.

adamabeshouse commented:
@fedde-s go ahead and squash. Let me know when it's ready and I'll take a final look, though it looks like you haven't changed much since I reviewed it. One thing is please use a more general method to HTML encode (e.g. https://stackoverflow.com/questions/14346414/how-do-you-do-html-encode-using-javascript)

fedde-s committed:
Prevent HTML injection in Oncoprint track tooltips

The description and title were only partially escaped, and then parsed
as HTML before inserting into the DOM element for the tooltip.
Interpreting them as plain text instead prevents unexpected behaviour
when IDs happen to contain HTML special characters.

fedde-s commented:
I squashed, and amended commits to
 1. use the `$().text('').html()` method to generate an HTML string in setup.js and
 2. not pass around dynamic HTML strings in the first place in oncoprintlabelview.js.

<details>
<summary>For comparison, these are the source code changes.</summary>

```Diff
diff --git a/portal/src/main/webapp/js/src/oncoprint/setup.js b/portal/src/main/webapp/js/src/oncoprint/setup.js
index 03ba46230..2841d0860 100644
--- a/portal/src/main/webapp/js/src/oncoprint/setup.js
+++ b/portal/src/main/webapp/js/src/oncoprint/setup.js
@@ -1345,12 +1345,10 @@ window.CreateCBioPortalOncoprintWithToolbar = function (ctr_selector, toolbar_se
 		for (i = 0; i < geneset_ids.length; i++) {
 		    track_geneset_id = geneset_ids[i];
 		    track_label_obj = new String(track_geneset_id);
-		    track_label_obj.html_content = (track_label_obj
-			    .replace(/&/g, '&amp;')
-			    .replace(/"/g, '&quot;')
-			    .replace(/'/g, '&#39;')
-			    .replace(/</g, '&lt;')
-			    .replace(/>/g, '&gt;')
+		    track_label_obj.html_content = (
+			    // encode the string as the textual contents of an
+			    // HTML element
+			    $('<div>').text(track_label_obj).html()
 			    // Include zero-width spaces to allow line breaks
 			    // after punctuation in (typically long) gs names
 			    .replace(/_/g, '_&#8203;')
diff --git a/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js b/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js
index 3439809e9..fd0143bba 100644
--- a/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js
+++ b/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js
@@ -60,13 +60,13 @@ var OncoprintLabelView = (function () {
 			var offset = view.$canvas.offset();   
 			if (isNecessaryToShortenLabel(view, view.labels[hovered_track])
 				|| view.track_link_urls[hovered_track]) {
-			    $tooltip_div.append($(formatTooltipHeader(
+			    $tooltip_div.append(formatTooltipHeader(
 				    view.labels[hovered_track],
-				    view.track_link_urls[hovered_track])));
+				    view.track_link_urls[hovered_track]));
 			}
-			var track_description = encodeHtml(view.track_descriptions[hovered_track]);
+			var track_description = view.track_descriptions[hovered_track];
 			if (track_description.length > 0) {
-			    $tooltip_div.append(track_description + "<br>");
+			    $tooltip_div.append($('<div>').text(track_description));
 			}
 			if (model.getContainingTrackGroup(hovered_track).length > 1) {
 			    view.$canvas.css('cursor', 'move');
@@ -148,28 +148,18 @@ var OncoprintLabelView = (function () {
 	    return label;
 	}
     };
-    var encodeHtml = function (text) {
-	return (text
-		.replace(/&/g, '&amp;')
-		.replace(/"/g, '&quot;')
-		.replace(/'/g, '&#39;')
-		.replace(/</g, '&lt;')
-		.replace(/>/g, '&gt;'));
-    };
     var formatTooltipHeader = function (label, link_url) {
-	var html_label = label.html_content || encodeHtml(label)
 	var header_contents;
 	if (link_url) {
 	    header_contents = (
-		    '<a href="' + encodeHtml(link_url) + '" target="_blank" rel="noopener noreferrer">'
-		    + html_label
-		    + '</a>');
+		    $('<a target="_blank" rel="noopener noreferrer">')
+		    .attr('href', link_url));
 	} else {
-	    header_contents = html_label;
+	    header_contents = $('<span>');
 	}
-	return ('<h5 style="font-weight: bold; margin: 0; font-size: inherit;">'
-		+ header_contents
-		+ '</h5>');
+	header_contents.append(label.html_content || document.createTextNode(label));
+	return ($('<h5 style="font-weight: bold; margin: 0; font-size: inherit;">')
+		.append(header_contents));
     };
     var renderAllLabels = function(view) {
 	if (view.rendering_suppressed) {
```

</summary>

adamabeshouse commented on portal/src/main/webapp/js/src/oncoprint/setup.js:
```diff
@@ -1338,12 +1338,21 @@ window.CreateCBioPortalOncoprintWithToolbar = function (ctr_selector, toolbar_se
+			    $('<div>').text(track_label_obj).html()
+			    // Include zero-width spaces to allow line breaks
+			    // after punctuation in (typically long) gs names
+			    .replace(/_/g, '_&#8203;')
```
these (1354-1355) still need to be here?

fedde-s replied:
Yes. The reason I am passing around HTML code is to insert zero-width spaces into the gene set ID (which can occasionally be a whole phrase in MSigDB's naming), indicating points at which the browser could split the ID over multiple lines instead of overflowing.

adamabeshouse replied:
interesting

adamabeshouse commented on portal/src/main/webapp/js/src/oncoprint/webgl/dist/oncoprint-bundle.js:
```diff
@@ -1656,14 +1657,17 @@ var OncoprintLabelView = (function () {
 		    var hovered_track = isMouseOnLabel(view, evt.pageY - view.$canvas.offset().top);
 		    if (hovered_track !== null) {
-			var $tooltip_div = $('<div>');
+			var $tooltip_div = $('<section>');
```
section?

fedde-s replied:
Considering what block element the boldfaced label should actually be semantically, I identified it as a heading of the tooltip, whatever the rank of that heading would be in the structure of the content (I guessed about 5). I can leave this commit out though, as the concatenation of label and description is no longer a problem now that I made the description itself a block element. Besides, the tooltip would have been an `<aside>` to the flow of content if anything.

fedde-s replied:
It makes no difference to browsers yet. What do you think, should I use an `<aside>`, or leave out the semantic annotations altogether for now, reverting it to the inline element `<b>`?

adamabeshouse replied:
I haven't seen this before - so they are basically the same as div but just supposed to signal a semantic difference to programmers?

fedde-s replied:
Ah, no, sorry, I should probably have referred to some documentation. I believe they were introduced in HTML5, along with the `<nav>`, `<article>`, `<header>` and `<footer>` tags, to signal additional semantic information to future browsers and alternative viewers (such as screen readers for visually impaired people). Here's one document about it: https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Using_HTML_sections_and_outlines

adamabeshouse replied:
oh, I see. but the difference is totally semantic, functionally its the same as a div? if so, then works for me

fedde-s commented:
Ah, no, sorry, I should probably have referred to some documentation. I believe they were introduced in HTML5, along with the `<nav>` tag, to signal additional semantic information to future browsers and alternative viewers (such as screen readers for visually handicapped people), information including outlines. Here's one document about it: https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Using_HTML_sections_and_outlines
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

pieterlukasse commented on web/src/main/java/org/cbioportal/web/GenesetController.java:
```diff
@@ -70,5 +72,16 @@
+    @RequestMapping(value = "/genesets/fetch", method = RequestMethod.POST,
+            consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ApiOperation("Fetch gene sets by ID")
+    public ResponseEntity<List<Geneset>> fetchGenes(
```
can we rename this to fetchGenesets?

pieterlukasse commented:
@fedde-s all seems good, so I would like to merge it. But I found one small copy+paste mistake you could still fix.

fedde-s commented:
I removed the commit that introduced new semantic HTML elements as I was unsure about whether they would indeed stay functionally equivalent, and amended the commit with the typo that Pieter found.

<details>
<summary>These are the resulting source code changes from what I proposed before.</summary>

```diff
diff --git a/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js b/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js
index fd0143bba..851d435e6 100644
--- a/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js
+++ b/portal/src/main/webapp/js/src/oncoprint/webgl/src/js/oncoprintlabelview.js
@@ -56,7 +56,7 @@ var OncoprintLabelView = (function () {
                } else {
                    var hovered_track = isMouseOnLabel(view, evt.pageY - view.$canvas.offset().top);
                    if (hovered_track !== null) {
-                       var $tooltip_div = $('<section>');
+                       var $tooltip_div = $('<div>');
                        var offset = view.$canvas.offset();
                        if (isNecessaryToShortenLabel(view, view.labels[hovered_track])
                                || view.track_link_urls[hovered_track]) {
@@ -158,8 +158,7 @@ var OncoprintLabelView = (function () {
            header_contents = $('<span>');
        }
        header_contents.append(label.html_content || document.createTextNode(label));
-       return ($('<h5 style="font-weight: bold; margin: 0; font-size: inherit;">')
-               .append(header_contents));
+       return $('<b style="display: block;">').append(header_contents);
     };
     var renderAllLabels = function(view) {
        if (view.rendering_suppressed) {
diff --git a/web/src/main/java/org/cbioportal/web/GenesetController.java b/web/src/main/java/org/cbioportal/web/GenesetController.java
index 5276eacea..7cf625da3 100644
--- a/web/src/main/java/org/cbioportal/web/GenesetController.java
+++ b/web/src/main/java/org/cbioportal/web/GenesetController.java
@@ -75,7 +75,7 @@ public class GenesetController {
     @RequestMapping(value = "/genesets/fetch", method = RequestMethod.POST,
             consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
     @ApiOperation("Fetch gene sets by ID")
-    public ResponseEntity<List<Geneset>> fetchGenes(
+    public ResponseEntity<List<Geneset>> fetchGenesets(
             @ApiParam(required = true, value = "List of gene set IDs")
             @Size(min = 1)
             @RequestBody List<String> genesetIds) {
```

</details>

fedde-s commented:
As far as I'm concerned, this is ready to merge now.
